mempool/reactor.go

```diff
@@ -100,19 +115,22 @@ func (memR *Reactor) Receive(e p2p.Envelope) {
 			memR.Logger.Error("received empty txs from peer", "src", e.Src)
 			return
 		}
-		txInfo := TxInfo{SenderID: memR.ids.GetForPeer(e.Src)}
-		if e.Src != nil {
-			txInfo.SenderP2PID = e.Src.ID()
-		}
 
 		var err error
-		for _, tx := range protoTxs {
-			ntx := types.Tx(tx)
-			err = memR.mempool.CheckTx(ntx, nil, txInfo)
+		for _, txBytes := range protoTxs {
+			tx := types.Tx(txBytes)
+
+			// Record the sender of every transaction received.
+			// Senders are stored until the transaction is removed from the mempool.
+			// Note that it's possible a tx is still in the cache but no longer in the mempool.
+			// For example, after committing a block, txs are removed from mempool but not the cache.
+			memR.addSender(tx.Key(), memR.ids.GetForPeer(e.Src))
```

Comment:
I think we should try to maintain the following invariant: 

  $tx \in memR.mempool.txs \iff tx \in memR.txSenders.keys()$. 

The rationale is to keep memory usage bounded. However, the code here is breaking such an invariant, because `tx` might not be added to the mempool.

Reply:
I agree and I have changed this code to record the sender only when the transaction is valid (and added to the mempool). Another problem with my previous code was that, if the received transaction was invalid, it would store the sender and never be removed.